src/utils/typedData.ts

```diff
@@ -107,130 +152,183 @@ function getMerkleTreeType(types: TypedData['types'], ctx: Context) {
   }
 
-  if (type === 'merkletree') {
-    const merkleTreeType = getMerkleTreeType(types, ctx);
-    const structHashes: string[] = (data as unknown[]).map((struct) => {
-      // eslint-disable-next-line @typescript-eslint/no-use-before-define
-      return encodeValue(types, merkleTreeType, struct as Record<string, unknown>)[1];
-    });
-    const { root } = new MerkleTree(structHashes as string[]);
-    return ['felt', root];
-  }
+  if (type.endsWith('*')) {
+    const isStructArray = Object.keys(types)
+      .map((x) => `${x}*`)
+      .includes(type);
 
-  if (type === 'felt*') {
-    return ['felt*', computeHashOnElements(data as string[])];
-  }
+    const mappingMethod: (entry: Record<string, unknown>) => string = isStructArray
+      ? (entry) => getStructHash(types, type.slice(0, -1), entry, revision)
+      : (entry) => encodeValue(types, type.slice(0, 1), entry, undefined, revision)[1];
 
-  if (type === 'selector') {
-    return ['felt', prepareSelector(data as string)];
+    const hashes: string[] = (data as Array<TypedData['message']>).map(mappingMethod);
+    return [type, revisionConfiguration[revision].hashMethod(hashes)];
   }
 
-  return [type, getHex(data as string)];
-};
+  switch (type) {
+    case 'enum': {
+      const [variantKey, variantData] = Object.entries(data as TypedData['message'])[0];
+
+      const parentType = types[ctx.parent as string][0] as StarkNetEnumType;
+      const enumType = types[parentType.contains];
+      const variantType = enumType.find((t) => t.name === variantKey) as StarkNetType;
+      const variantIndex = enumType.indexOf(variantType);
+
+      const encodedSubtypes = variantType.type
+        .slice(1, -1)
+        .split(',')
+        .map((subtype, index) => {
+          const subtypeData = (variantData as unknown[])[index];
+          return encodeValue(types, subtype, subtypeData, undefined, revision)[1];
+        });
+      return [type, revisionConfiguration[revision].hashMethod([variantIndex, ...encodedSubtypes])];
+    }
+    case 'merkletree': {
+      const merkleTreeType = getMerkleTreeType(types, ctx);
+      const structHashes: string[] = (data as Array<TypedData['message']>).map((struct) => {
+        return encodeValue(types, merkleTreeType, struct, undefined, revision)[1];
+      });
+      const { root } = new MerkleTree(structHashes as string[]);
+      return ['felt', root];
+    }
+    case 'selector': {
+      return ['felt', prepareSelector(data as string)];
+    }
+    case 'string': {
+      if (revision === Revision.Active) {
+        // TODO: should this be skipped for v5 and added as ByteArray in v6?
+        const elements = splitLongString(data as string).map(getHex);
+        return [type, revisionConfiguration[revision].hashMethod(elements)];
+      }
+      return [type, getHex(data as string)];
+    }
+    default: {
+      return [type, getHex(data as string)];
+    }
```

Comment:
this is accepting many types that are not defined in the spec, it also allows all revision > 0 types for revision 0

Reply:
also some types like shortstring need special treatment (length check)